File: src/test/java/com/testainers/BasicAuthResourceTest.java
package com.testainers;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.Method;
import jakarta.ws.rs.core.HttpHeaders;
import org.junit.jupiter.api.Test;

import static org.hamcrest.Matchers.*;

/**
 * @author Eduardo Folly
 */
@QuarkusTest
public class BasicAuthResourceTest extends BaseResourceTest {

    private static final String user = "test";
    private static final String pass = "test-pass0";

    @Test
    public void noHeader() {
        for (Method method : METHODS) {
            json(method)
                    .pathParam("user", user)
                    .pathParam("pass", pass)
                    .request(method, "/basic-auth/{user}/{pass}")
                    .then()
                    .statusCode(401)
                    .body("body.body",
                          method == Method.GET ? nullValue() : is(BODY),
                          bodyMatchers(
                                  method,
                                  "body.auth", is(false),
                                  "body.user", is(user),
                                  "body.pass", is(pass),
                                  "body.message",
                                  is("Authorization header not present.")
                          ));
        }
    }

    @Test
    public void emptyHeader() {
        for (Method method : METHODS) {
            json(method)
                    .header(HttpHeaders.AUTHORIZATION, "")
                    .pathParam("user", user)
                    .pathParam("pass", pass)
                    .request(method, "/basic-auth/{user}/{pass}")
                    .then()
                    .statusCode(401)
                    .body("body.body",
                          method == Method.GET ? nullValue() : is(BODY),
                          bodyMatchers(
                                  method,
                                  "body.auth", is(false),
                                  "body.user", is(user),
                                  "body.pass", is(pass),
                                  "body.message",
                                  is("Authorization header not present.")
                          ));
        }
    }

    @Test
    public void success() {
        for (Method method : METHODS) {
            json(method)
                    .auth().preemptive().basic(user, pass)
                    .pathParam("user", user)
                    .pathParam("pass", pass)
                    .request(method, "/basic-auth/{user}/{pass}")
                    .then()
                    .statusCode(200)
                    .body("body.body",
                          method == Method.GET ? nullValue() : is(BODY),
                          bodyMatchers(
                                  method,
                                  "body.auth", is(true),
                                  "body.user", is(user),
                                  "body.pass", is(pass),
                                  "body.message", is("Success.")
                          ));
        }
    }

    @Test
    public void userFail() {
        for (Method method : METHODS) {
            json(method)
                    .auth().preemptive().basic(user, pass)
                    .pathParam("user", user + "Fail")
                    .pathParam("pass", pass)
                    .request(method, "/basic-auth/{user}/{pass}")
                    .then()
                    .statusCode(403)
                    .body("body.body",
                          method == Method.GET ? nullValue() : is(BODY),
                          bodyMatchers(
                                  method,
                                  "body.auth", is(false),
                                  "body.user", is(user + "Fail"),
                                  "body.pass", is(pass),
                                  "body.message", is("Forbidden.")
                          ));
        }
    }

    @Test
    public void passFail() {
        for (Method method : METHODS) {
            json(method)
                    .auth().preemptive().basic(user, pass)
                    .pathParam("user", user)
                    .pathParam("pass", pass + "Fail")
                    .request(method, "/basic-auth/{user}/{pass}")
                    .then()
                    .statusCode(403)
                    .body("body.body",
                          method == Method.GET ? nullValue() : is(BODY),
                          bodyMatchers(
                                  method,
                                  "body.auth", is(false),
                                  "body.user", is(user),
                                  "body.pass", is(pass + "Fail"),
                                  "body.message", is("Forbidden.")
                          ));
        }
    }

}

File: src/test/kotlin/com/testainers/BasicAuthResourceTest.kt
package com.testainers

import io.quarkus.test.junit.QuarkusTest
import io.restassured.http.Method
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.nullValue
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

/**
 * @author Eduardo Folly
 */
@QuarkusTest
class BasicAuthResourceTest : BaseResourceTest() {
    val user = "test"
    val pass = "test-pass0"

    @ParameterizedTest
    @MethodSource("onlyMethods")
    fun noHeader(method: Method) {
        json(method)
            .request(method, "/basic-auth/$user/$pass")
            .then()
            .statusCode(401)
            .apply {
                if (method != Method.HEAD) {
                    body(
                        "body.body",
                        if (method == Method.GET) {
                            nullValue()
                        } else {
                            equalTo(body)
                        },
                        *bodyMatchers(
                            method,
                            mapOf(
                                "body.auth" to equalTo(false),
                                "body.user" to equalTo(user),
                                "body.pass" to equalTo(pass),
                                "body.message" to
                                    equalTo(
                                        "Authorization header not present.",
                                    ),
                            ),
                        ),
                    )
                }
            }
    }

    @ParameterizedTest
    @MethodSource("onlyMethods")
    fun success(method: Method) {
        json(method)
            .auth()
            .preemptive()
            .basic(user, pass)
            .request(method, "/basic-auth/$user/$pass")
            .then()
            .statusCode(200)
            .apply {
                if (method != Method.HEAD) {
                    body(
                        "body.body",
                        if (method == Method.GET) {
                            nullValue()
                        } else {
                            equalTo(body)
                        },
                        *bodyMatchers(
                            method,
                            mapOf(
                                "body.auth" to equalTo(true),
                                "body.user" to equalTo(user),
                                "body.pass" to equalTo(pass),
                                "body.message" to equalTo("Success."),
                            ),
                        ),
                    )
                }
            }
    }

    @ParameterizedTest
    @MethodSource("onlyMethods")
    fun fail(method: Method) {
        json(method)
            .auth()
            .preemptive()
            .basic(user + "A", pass + "A")
            .request(method, "/basic-auth/$user/$pass")
            .then()
            .statusCode(403)
            .apply {
                if (method != Method.HEAD) {
                    body(
                        "body.body",
                        if (method == Method.GET) {
                            nullValue()
                        } else {
                            equalTo(body)
                        },
                        *bodyMatchers(
                            method,
                            mapOf(
                                "body.auth" to equalTo(false),
                                "body.user" to equalTo(user),
                                "body.pass" to equalTo(pass),
                                "body.message" to equalTo("Forbidden."),
                            ),
                        ),
                    )
                }
            }
    }
}
